cpp/fanos.cc

```diff
@@ -0,0 +1,55 @@
+// fanos.cc
+
+#include "cpp/fanos.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "cpp/fanos_shared.h"
+
+namespace fanos {
+
+static constexpr int kNumFds = 3;
+static constexpr int kSizeOfFds = sizeof(int) * kNumFds;
+
+void send(int sock_fd, Str* blob) {
+  int fds[kNumFds] = {-1, -1, -1};
+  char* value_err = NULL;
+  int err_code = 0;
+  fanos_send(sock_fd, blob->data(), len(blob), fds, &err_code, &value_err);
+  if (err_code != 0) {
+    throw Alloc<IOError>(err_code);
+  }
+  if (value_err != NULL) {
+    throw Alloc<ValueError>(StrFromC(value_err));
+  }
+}
+
+Str* recv(int sock_fd, List<int>* fd_out) {
+  int fds[kNumFds] = {-1, -1, -1};
+  char* value_err = NULL;
+  int err_code = 0;
+  int data_len = -1;
+  char* data_buf = fanos_recv(sock_fd, fds, &data_len, &err_code, &value_err);
```

Comment:
Oh I think data_buf might leak here?  Because it is malloc'd in fanos_recv.  

StrFromC() is always a copy -- the copy becomes owned by the GC.  But the original still exists


Since fanos_shared.c has two callers, I think it's a good idea to document the memory management responsibilities in the interface,  I guess in the header file

---

Also just like the struct for errors, it might be useful to have

```
struct FanosResult {
  char* data;
  int len;
}
```

and call the same way:

```
FanosResult result = { 0 };
FanosError e = { 0 };

if (fanos_recv(sock_fd, fds, &result, &e)) {  // success or failure
  Str* ret = StrFromC(result->data, result->len);
  free(result->data);  
  ...
} else {
  // Handle errors
}
```


I tend to use the "out params" style a lot in C, since it is pretty general and functions written in the style compose pretty easily

Reply:
I suppose errno could also be the return value, so you check

`if (fanos_recv(...) != 0)`

I guess whatever makes the call stack more consistent / shorter

Reply:
Passing the error struct as an output param felt slightly cleaner to me